### PR TITLE
Signature algorithm refactor part 2.

### PR DESCRIPTION
**Pull request by snhenson**
##### Description of change

This PR is the second part of the signature algorithms reorganisation.

The digest and public key indices are added to the SIGALG_LOOKUP table. This simplifies several pieces of code which no longer need to lookup the values in a separate table.

The peer signature algorithm used is stored instead of separate digest and signature type fields.

The test for ciphersuite and certificate type consistency has been removed for TLS 1.3 as the ciphersuites can be used with all certificate types. This means we can now interop with picotls server and ECDSA. We can't test that (yet) because server side ECDSA and TLS 1.3 is not currently supported.

### CODE REVIEW

File: ssl/ssl_locl.h [hunk 1]
```diff
@@ -1180,6 +1186,30 @@ struct ssl_st {
     CRYPTO_RWLOCK *lock;
 };
 
+/*
+ * Structure containing table entry of values associated with the signature
+ * algorithms (signature scheme) extension
+*/
+typedef struct sigalg_lookup_st {
+    /* TLS 1.3 signature scheme name */
```
Comment (richsalz) [inline ssl/ssl_locl.h]: is it reasonable to put the comment on same line as the field?  okay if not.

File: ssl/ssl_locl.h [hunk 2]
```diff
@@ -1180,6 +1186,30 @@ struct ssl_st {
...
+    /* NID of hash algorithm */
+    int hash;
+    /* Index of hash algorithm */
+    int hash_idx;
+    /* NID of signature algorithm */
+    int sig;
+    /* Index of signature algorithm */
+    int sig_idx;
+    /* NID of signature algorithm */
+    /* Combined hash and signature NID, if any */
```
Comment (richsalz) [inline ssl/ssl_locl.h]: Is this second comment the definition of what the sigalg is?  if not, I'm confused.

File: ssl/ssl_locl.h [hunk 3]
```diff
@@ -1269,10 +1299,8 @@ typedef struct ssl3_state_st {
         uint16_t *peer_sigalgs;
         /* Size of above array */
         size_t peer_sigalgslen;
-        /* Digest peer uses for signing */
-        const EVP_MD *peer_md;
-        /* Signature type: public key type or EVP_PKEY_RSA_PSS for PSS */
-        int peer_sigtype;
+        /* Sigalg peer actualy uses */
```
Comment (richsalz) [inline ssl/ssl_locl.h]: actually :)

File: ssl/statem/statem_clnt.c [hunk 4]
```diff
@@ -1562,17 +1562,23 @@ MSG_PROCESS_RETURN tls_process_server_certificate(SSL *s, PACKET *pkt)
...
-            (i != SSL_PKEY_GOST12_512 && i != SSL_PKEY_GOST12_256
-             && i != SSL_PKEY_GOST01))) {
-        x = NULL;
-        al = SSL_AD_ILLEGAL_PARAMETER;
-        SSLerr(SSL_F_TLS_PROCESS_SERVER_CERTIFICATE,
-               SSL_R_WRONG_CERTIFICATE_TYPE);
-        goto f_err;
+    /*
+     * Check certificate type is consistent with ciphersuite. For TLS 1.3
+     * skip check as TLS 1.3 ciphersuites can be used with any certificate
```
Comment (richsalz) [inline ssl/statem/statem_clnt.c]: perhaps "skip check, since" ?

File: ssl/statem/statem_clnt.c [hunk 5]
```diff
@@ -1562,17 +1562,23 @@ MSG_PROCESS_RETURN tls_process_server_certificate(SSL *s, PACKET *pkt)
...
+     * type.
+     */
+    if (!SSL_IS_TLS13(s)) {
+        exp_idx = ssl_cipher_get_cert_index(s->s3->tmp.new_cipher);
+        if (exp_idx >= 0 && i != exp_idx
+            && (exp_idx != SSL_PKEY_GOST_EC ||
+                (i != SSL_PKEY_GOST12_512 && i != SSL_PKEY_GOST12_256
```
Comment (richsalz) [inline ssl/statem/statem_clnt.c]: maaybe a define for ISNT_GOST makes sense?  or !IS_GOST ?

File: ssl/t1_lib.c [hunk 6]
```diff
@@ -823,57 +835,74 @@ size_t tls12_get_psigalgs(SSL *s, int sent, const uint16_t **psigs)
...
-        if (!tls1_set_ec_id(curve_id, &comp_id, EVP_PKEY_get0_EC_KEY(pkey)))
-            return 0;
-        if (!s->server && !tls1_check_ec_key(s, curve_id, &comp_id)) {
-            SSLerr(SSL_F_TLS12_CHECK_PEER_SIGALG, SSL_R_WRONG_CURVE);
-            return 0;
-        }
-        /* If Suite B only P-384+SHA384 or P-256+SHA-256 allowed */
-        if (tls1_suiteb(s)) {
-            if (curve_id[0])
+        EC_KEY *ec = EVP_PKEY_get0_EC_KEY(pkey);
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after decl.

File: ssl/t1_lib.c [hunk 7]
```diff
@@ -823,57 +835,74 @@ size_t tls12_get_psigalgs(SSL *s, int sent, const uint16_t **psigs)
...
+        if (SSL_IS_TLS13(s)) {
+            /* For TLS 1.3 check curve matches signature algorithm */
+            int curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec));
+            if (curve != lu->curve) {
+                SSLerr(SSL_F_TLS12_CHECK_PEER_SIGALG, SSL_R_WRONG_CURVE);
+                return 0;
+            }
+        } else {
+            unsigned char curve_id[2], comp_id;
```
Comment (richsalz) [inline ssl/t1_lib.c]: again blank line.

File: ssl/t1_lib.c [hunk 8]
```diff
@@ -823,57 +835,74 @@ size_t tls12_get_psigalgs(SSL *s, int sent, const uint16_t **psigs)
...
-            SSLerr(SSL_F_TLS12_CHECK_PEER_SIGALG, SSL_R_WRONG_CURVE);
-            return 0;
-        }
-        /* If Suite B only P-384+SHA384 or P-256+SHA-256 allowed */
-        if (tls1_suiteb(s)) {
-            if (curve_id[0])
+        EC_KEY *ec = EVP_PKEY_get0_EC_KEY(pkey);
+        if (SSL_IS_TLS13(s)) {
+            /* For TLS 1.3 check curve matches signature algorithm */
+            int curve = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec));
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line.

File: ssl/t1_lib.c [hunk 9]
```diff
@@ -823,57 +835,74 @@ size_t tls12_get_psigalgs(SSL *s, int sent, const uint16_t **psigs)
...
+            /* If Suite B only P-384+SHA384 or P-256+SHA-256 allowed */
+            if (tls1_suiteb(s)) {
+                if (curve_id[0])
                     return 0;
-                }
-            } else if (curve_id[1] == TLSEXT_curve_P_384) {
-                if (tls_sigalg_get_hash(sig) != NID_sha384) {
-                    SSLerr(SSL_F_TLS12_CHECK_PEER_SIGALG,
-                           SSL_R_ILLEGAL_SUITEB_DIGEST);
+                if (curve_id[1] == TLSEXT_curve_P_256) {
```
Comment (richsalz) [inline ssl/t1_lib.c]: boy, that's a nest of if's!  I can't think of a cleaner way to do it, tho.

File: ssl/t1_lib.c [hunk 10]
```diff
@@ -2002,8 +1985,8 @@ int tls1_check_chain(SSL *s, X509 *x, EVP_PKEY *pk, STACK_OF(X509) *chain,
             size_t j;
             const uint16_t *p = c->conf_sigalgs;
             for (j = 0; j < c->conf_sigalgslen; j++, p++) {
-                if (tls_sigalg_get_hash(*p) == NID_sha1
-                        && tls_sigalg_get_sig(*p) == rsign)
+                const SIGALG_LOOKUP *lu = tls1_lookup_sigalg(*p);
+                if (lu && lu->hash == NID_sha1 && lu->sig == rsign)
```
Comment (richsalz) [inline ssl/t1_lib.c]: blank line after the decl.  add explicit "!= NULL" test

File: ssl/t1_lib.c (hunk 9 above)
Comment (levitte) [inline ssl/t1_lib.c]: A switch?

``` C
            switch (curve_id[1]) {
            case TLSEXT_CURVE_P_256:
                ...
                break;
            case TLSEXT_CURVE_P_384:
                ...
                break;
            default:
                return 0;
            }
```

File: ssl/ssl_locl.h (hunk 2 above)
Comment (snhenson) [inline ssl/ssl_locl.h]: Oops, added another comment for some reason, removed.

File: ssl/statem/statem_clnt.c (hunk 4 above)
Comment (snhenson) [inline ssl/statem/statem_clnt.c]: Done.

File: ssl/t1_lib.c (hunk 9 above)
Comment (snhenson) [inline ssl/t1_lib.c]: That's just the original "logic". There might be a cleaner way using the new tables, I'll have a look.

Comment (snhenson) [inline ssl/t1_lib.c]: Added a new commit that is functionally equivalent to the original but tidier.

**Comment by snhenson** [comment]
Ping @richsalz I think this has addressed your comments now. Please check.

**Comment by snhenson** [comment]
Thanks, pushed.

**Pull request closed without merging**